store/fifo_store.go

```diff
@@ -43,30 +45,52 @@ func (f *FIFOStore[K, V]) Get(key K) (V, bool) {
 
 // Set
 func (f *FIFOStore[K, V]) Set(key K, value V) {
-	if _, exists := f.Store.Get(key); exists {
-		return
-	}
+	_, exists := f.Store.Get(key)
 
 	// Remove older
-	if f.order != nil && f.order.Len() >= f.capacity {
+	if f.order != nil && f.order.Len() >= f.capacity && !exists {
 		oldest := f.order.Front()
 		if oldest != nil {
+			f.lock.Lock()
 			f.order.Remove(oldest)
+			f.lock.Unlock()
 			f.Store.Remove(oldest.Value.(K))
 		}
 	}
 
+	f.lock.Lock()
 	f.order.PushBack(key)
+	f.lock.Unlock()
 	f.Store.Set(key, value)
 }
```

Comment:
A question on the behavior of `Set`:

If an element exists and you call `Set()` on that element, what should happen to the `order` of that element? Right now it looks like it will save the order twice in `f.order`. e.g.
```
fifo.Set("color", "white")
fifo.Set("color", "white")
fifo.order.Len() <-- 2
len(fifo.Store.Keys()) <-- 1
```
I don't think that's correct because then you'll have a mismatch between the elements in `f.Store` and `f.order`. If the same element is accessed many times `f.order` could be full of just one element.

To hopefully keep things simple, I think what we want when the key is found is to update the value of the key in `f.Store` but leave the order unchanged. You'll probably want to first check the capacity and remove the oldest key before checking if the key exists or not. Also to make things simpler, you should be able to `lock/unlock` once at the beginning of this method with `defer`:
```
f.lock.Lock()
defer f.lock.Unlock()
```
That might also help prevent race conditions where `f.Store` gets updated between calling `f.Store.Get()` and `f.Store.Set()`.

Reply:
That makes sense!